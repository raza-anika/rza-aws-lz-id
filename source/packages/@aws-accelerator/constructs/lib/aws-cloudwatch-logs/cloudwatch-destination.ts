/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Construction properties for CloudWatch Destination for Kinesis Stream.
 */
export interface CloudWatchDestinationProps {
  /**
   *
   * KMS key to encrypt the Kinesis Data Stream
   */
  kinesisKmsKey: cdk.aws_kms.IKey;
  /**
   *
   * Kinesis Data Stream for CloudWatch logs
   */
  kinesisStream: cdk.aws_kinesis.IStream;
  /**
   * Organization ID to restrict the usage within specific org
   */
  organizationId?: string;
  /**
   * Partition to determine the IAM condition.
   */
  partition: string;
  /**
   * Account IDs for the IAM condition.
   */
  accountIds?: string[];
}
/**
 * Class to configure CloudWatch Destination on logs receiving account
 */
export class CloudWatchDestination extends Construct {
  constructor(scope: Construct, id: string, props: CloudWatchDestinationProps) {
    super(scope, id);

    //Create policy for access to Kinesis stream
    const kinesisStreamAccess = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: ['kinesis:ListShards', 'kinesis:PutRecord', 'kinesis:PutRecords'],
          resources: [props.kinesisStream.streamArn],
        }),
      ],
    });
    const kmsKeyAccess = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: [
            'kms:Decrypt',
            'kms:Encrypt',
            'kms:GenerateDataKey',
            'kms:ReEncryptTo',
            'kms:GenerateDataKeyWithoutPlaintext',
            'kms:GenerateDataKeyPairWithoutPlaintext',
            'kms:GenerateDataKeyPair',
            'kms:ReEncryptFrom',
          ],
          resources: [props.kinesisKmsKey.keyArn],
        }),
      ],
    });

    // Create a role for CloudWatch Logs destination
    const logsKinesisRole = new cdk.aws_iam.Role(this, 'LogsKinesisRole', {
      assumedBy: new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.${cdk.Stack.of(this).urlSuffix}`),
      // this needs to be inline to ensure role is created with proper access
      // if not, CloudWatch destination creation fails with no permission to access Kinesis or KMS (generateDataKey access error)
      inlinePolicies: {
        KinesisAccess: kinesisStreamAccess,
        KmsAccess: kmsKeyAccess,
      },
    });

    let principalOrgIdCondition: object | undefined = undefined;
    let accountPrincipals: object | cdk.aws_iam.IPrincipal;

    if (props.partition === 'aws-cn' || !props.organizationId) {
      // Only principal block with list of account id is supported.
      accountPrincipals = {
        AWS: props.accountIds,
      };
    } else {
      principalOrgIdCondition = {
        StringEquals: {
          'aws:PrincipalOrgID': props.organizationId,
        },
      };
      accountPrincipals = new cdk.aws_iam.AnyPrincipal();
    }

    const logDestinationPolicy = JSON.stringify({
      Version: '2012-10-17',
      Statement: [
        {
          Effect: 'Allow',
          Principal: accountPrincipals,
          Action: 'logs:PutSubscriptionFilter',
          Resource: `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:destination:AWSAcceleratorCloudWatchToS3`,
          Condition: principalOrgIdCondition,
        },
      ],
    });

    new cdk.aws_logs.CfnDestination(this, 'Resource', {
      roleArn: logsKinesisRole.roleArn,
      targetArn: props.kinesisStream.streamArn,
      destinationName: 'AWSAcceleratorCloudWatchToS3',
      destinationPolicy: logDestinationPolicy,
    });
  }
}
