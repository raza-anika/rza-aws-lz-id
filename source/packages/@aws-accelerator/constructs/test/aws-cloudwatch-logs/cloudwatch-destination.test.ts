/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { CloudWatchDestination } from '../../lib/aws-cloudwatch-logs/cloudwatch-destination';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(CloudWatchDestination): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();
const cnStack = new cdk.Stack();

new CloudWatchDestination(stack, 'CloudWatchDestination', {
  kinesisKmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  kinesisStream: new cdk.aws_kinesis.Stream(stack, 'CustomStream', {}),
  organizationId: 'o-some-org-id',
  partition: 'aws',
});

new CloudWatchDestination(cnStack, 'CloudWatchDestination', {
  kinesisKmsKey: new cdk.aws_kms.Key(cnStack, 'CustomKey', {}),
  kinesisStream: new cdk.aws_kinesis.Stream(cnStack, 'CustomStream', {}),
  organizationId: 'o-some-org-id',
  accountIds: ['111111111111', '222222222222'],
  partition: 'aws-cn',
});

/**
 * CloudWatchDestination construct test
 */
describe('CloudWatchDestination', () => {
  snapShotTest(testNamePrefix, stack);
  snapShotTest(testNamePrefix, cnStack);
});
