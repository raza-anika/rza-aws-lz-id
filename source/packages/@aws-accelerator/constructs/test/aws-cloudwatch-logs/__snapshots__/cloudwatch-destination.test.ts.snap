// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudWatchDestination Construct(CloudWatchDestination):  Snapshot Test 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": Object {
    "CloudWatchDestination47668810": Object {
      "Properties": Object {
        "DestinationName": "AWSAcceleratorCloudWatchToS3",
        "DestinationPolicy": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"AWS\\":[\\"*\\"]},\\"Action\\":\\"logs:PutSubscriptionFilter\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":destination:AWSAcceleratorCloudWatchToS3\\",\\"Condition\\":{\\"StringEquals\\":{\\"aws:PrincipalOrgID\\":\\"o-some-org-id\\"}}}]}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "CloudWatchDestinationLogsKinesisRole54686795",
            "Arn",
          ],
        },
        "TargetArn": Object {
          "Fn::GetAtt": Array [
            "CustomStreamE8E9158E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Logs::Destination",
    },
    "CloudWatchDestinationLogsKinesisRole54686795": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "logs.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesis:ListShards",
                    "kinesis:PutRecord",
                    "kinesis:PutRecords",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomStreamE8E9158E",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisAccess",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:GenerateDataKey",
                    "kms:ReEncryptTo",
                    "kms:GenerateDataKeyWithoutPlaintext",
                    "kms:GenerateDataKeyPairWithoutPlaintext",
                    "kms:GenerateDataKeyPair",
                    "kms:ReEncryptFrom",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomKey1E6D0D07",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KmsAccess",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomKey1E6D0D07": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomStreamE8E9158E": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;

exports[`CloudWatchDestination Construct(CloudWatchDestination):  Snapshot Test 2`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": Object {
    "CloudWatchDestination47668810": Object {
      "Properties": Object {
        "DestinationName": "AWSAcceleratorCloudWatchToS3",
        "DestinationPolicy": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"AWS\\":[\\"111111111111\\",\\"222222222222\\"]},\\"Action\\":\\"logs:PutSubscriptionFilter\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":destination:AWSAcceleratorCloudWatchToS3\\"}]}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "CloudWatchDestinationLogsKinesisRole54686795",
            "Arn",
          ],
        },
        "TargetArn": Object {
          "Fn::GetAtt": Array [
            "CustomStreamE8E9158E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Logs::Destination",
    },
    "CloudWatchDestinationLogsKinesisRole54686795": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "logs.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesis:ListShards",
                    "kinesis:PutRecord",
                    "kinesis:PutRecords",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomStreamE8E9158E",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisAccess",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:GenerateDataKey",
                    "kms:ReEncryptTo",
                    "kms:GenerateDataKeyWithoutPlaintext",
                    "kms:GenerateDataKeyPairWithoutPlaintext",
                    "kms:GenerateDataKeyPair",
                    "kms:ReEncryptFrom",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomKey1E6D0D07",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KmsAccess",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomKey1E6D0D07": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomStreamE8E9158E": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
